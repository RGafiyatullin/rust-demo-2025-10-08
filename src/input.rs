//! Data types to process input: transaction and its parts.

use crate::types::{ClientId, PositiveAmount, TxId};

/// A transaction of any supported kind.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct Tx {
    /// Client this transaction relates to.
    #[serde(rename = "client")]
    pub client_id: ClientId,

    /// Globally unique transaction id.
    #[serde(rename = "tx")]
    pub tx_id: TxId,

    /// See [`TxKind`].
    #[serde(flatten)]
    pub kind: TxKind,
}

/// Data specific to the transaction kind.
#[derive(Debug, Clone, serde::Deserialize)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum TxKind {
    /// See [`TxDeposit`].
    Deposit(TxDeposit),
    /// See [`TxWithdrawal`].
    Withdrawal(TxWithdrawal),
    /// Initiate a dispute.
    Dispute,
    /// Cancel the previously raised dispute: unhold the disputed funds.
    Resolve,
    /// Withdraw the disputed funds.
    Chargeback,
}

/// Put funds into the account.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct TxDeposit {
    /// Amount being deposited
    #[serde(rename = "amount")]
    pub amount_deposited: PositiveAmount,
}

/// Take funds from the account.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct TxWithdrawal {
    /// Amount to withdraw
    #[serde(rename = "amount")]
    pub amount_withdrawn: PositiveAmount,
}

#[cfg(test)]
mod tests;
