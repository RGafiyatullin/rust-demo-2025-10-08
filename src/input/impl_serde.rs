//! This module contains a separate set of structures do deserialize
//! transactions.
//!
//! The necessity for such seemingly redundant approach emerged due to the way
//! `serde(flatten)` works: it forces `deserialize_any`-path, thus eagerly
//! treats as floats/integers those fields, whose data looks like such. Treating
//! decimal values as floats can cause precision problems, it generally should
//! be avoided.
//!
//! IDDQD: https://chatgpt.com/share/68e52644-bea4-800f-ae3e-47cec9dbbb66

use serde::Deserialize;

use crate::{
    input::{Tx, TxDeposit, TxKind, TxWithdrawal},
    types::{ClientId, PositiveAmount, TxId},
};

#[derive(serde::Deserialize)]
struct T {
    #[serde(rename = "type")]
    kind: K,
    #[serde(rename = "client")]
    client_id: ClientId,
    #[serde(rename = "tx")]
    tx_id: TxId,
    #[serde(rename = "amount")]
    amount_opt: Option<PositiveAmount>,
}

#[derive(serde::Deserialize)]
#[serde(rename_all = "lowercase")]
enum K {
    Deposit,
    Withdrawal,
    Dispute,
    Resolve,
    Chargeback,
}

impl<'de> Deserialize<'de> for Tx {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        use serde::de::Error as _;

        let T {
            kind,
            client_id,
            tx_id,
            amount_opt: amount,
        } = Deserialize::deserialize(deserializer)?;

        let kind = match (kind, amount) {
            (K::Deposit, Some(amount_deposited)) => TxKind::Deposit(TxDeposit { amount_deposited }),
            (K::Withdrawal, Some(amount_withdrawn)) => {
                TxKind::Withdrawal(TxWithdrawal { amount_withdrawn })
            }
            (K::Deposit | K::Withdrawal, None) => {
                Err(D::Error::custom("field `amount` is missing"))?
            }
            (K::Dispute, _) => TxKind::Dispute,
            (K::Resolve, _) => TxKind::Resolve,
            (K::Chargeback, _) => TxKind::Chargeback,
        };

        Ok(Self {
            client_id,
            tx_id,
            kind,
        })
    }
}
