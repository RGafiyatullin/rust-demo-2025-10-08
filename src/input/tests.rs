use std::path::Path;

use test_case::test_case;

use crate::input::Tx;

#[test_case("deposits")]
#[test_case("withdrawals")]
#[test_case("disputes")]
#[test_case("resolves")]
#[test_case("chargebacks")]
fn parse_csv(case_name: &str) {
    let input_file = Path::new(file!())
        .parent()
        .expect("file!().parent")
        .join("cases")
        .join(format!("{}.csv", case_name));
    let mut output = vec![];
    let csv_reader = csv::ReaderBuilder::new()
        .flexible(true)
        .trim(csv::Trim::All)
        .from_path(input_file)
        .expect("CsvReader::from_path");
    for decode_result in csv_reader.into_deserialize::<Tx>() {
        output.push(decode_result.map_err(|e| e.to_string()));
    }
    insta::with_settings!({
        snapshot_path => "cases",
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_debug_snapshot!(case_name, output);
    });
}
