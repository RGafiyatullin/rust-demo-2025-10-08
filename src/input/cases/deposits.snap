---
source: src/input/tests.rs
expression: output
---
[
    Ok(
        Tx {
            client_id: ClientId(
                1,
            ),
            tx_id: TxId(
                1,
            ),
            kind: Deposit(
                TxDeposit {
                    amount_deposited: PositiveAmount(
                        1.0,
                    ),
                },
            ),
        },
    ),
    Ok(
        Tx {
            client_id: ClientId(
                1,
            ),
            tx_id: TxId(
                2,
            ),
            kind: Deposit(
                TxDeposit {
                    amount_deposited: PositiveAmount(
                        0.0,
                    ),
                },
            ),
        },
    ),
    Ok(
        Tx {
            client_id: ClientId(
                1,
            ),
            tx_id: TxId(
                3,
            ),
            kind: Deposit(
                TxDeposit {
                    amount_deposited: PositiveAmount(
                        -1.0,
                    ),
                },
            ),
        },
    ),
    Ok(
        Tx {
            client_id: ClientId(
                2,
            ),
            tx_id: TxId(
                4,
            ),
            kind: Deposit(
                TxDeposit {
                    amount_deposited: PositiveAmount(
                        1.0,
                    ),
                },
            ),
        },
    ),
    Ok(
        Tx {
            client_id: ClientId(
                2,
            ),
            tx_id: TxId(
                5,
            ),
            kind: Deposit(
                TxDeposit {
                    amount_deposited: PositiveAmount(
                        0.0,
                    ),
                },
            ),
        },
    ),
    Ok(
        Tx {
            client_id: ClientId(
                2,
            ),
            tx_id: TxId(
                6,
            ),
            kind: Deposit(
                TxDeposit {
                    amount_deposited: PositiveAmount(
                        -1.0,
                    ),
                },
            ),
        },
    ),
    Ok(
        Tx {
            client_id: ClientId(
                3,
            ),
            tx_id: TxId(
                7,
            ),
            kind: Deposit(
                TxDeposit {
                    amount_deposited: PositiveAmount(
                        1.0,
                    ),
                },
            ),
        },
    ),
    Ok(
        Tx {
            client_id: ClientId(
                3,
            ),
            tx_id: TxId(
                8,
            ),
            kind: Deposit(
                TxDeposit {
                    amount_deposited: PositiveAmount(
                        1.0001,
                    ),
                },
            ),
        },
    ),
    Ok(
        Tx {
            client_id: ClientId(
                3,
            ),
            tx_id: TxId(
                9,
            ),
            kind: Deposit(
                TxDeposit {
                    amount_deposited: PositiveAmount(
                        1.0001,
                    ),
                },
            ),
        },
    ),
    Ok(
        Tx {
            client_id: ClientId(
                3,
            ),
            tx_id: TxId(
                10,
            ),
            kind: Deposit(
                TxDeposit {
                    amount_deposited: PositiveAmount(
                        1.0001,
                    ),
                },
            ),
        },
    ),
    Ok(
        Tx {
            client_id: ClientId(
                4,
            ),
            tx_id: TxId(
                11,
            ),
            kind: Deposit(
                TxDeposit {
                    amount_deposited: PositiveAmount(
                        111111111111111.1,
                    ),
                },
            ),
        },
    ),
    Err(
        Error(
            Deserialize {
                pos: Some(
                    Position {
                        byte: 366,
                        line: 13,
                        record: 12,
                    },
                ),
                err: DeserializeError {
                    field: None,
                    kind: Message(
                        "invalid value: floating point `1111111111111111.1`, expected number or string containing a fixed-point number",
                    ),
                },
            },
        ),
    ),
]
